File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
void solve(int idx, int n , vector<int>&candidates, vector<vector<int>>&ans, vector<int>temp, int target){
        if(idx==n){
            if(target == 0){
                ans.push_back(temp);
                return;
            }
            return;
        }
        temp.push_back(candidates[idx]);
       if(candidates[idx] <= target) solve(idx,n,candidates,ans,temp,target - candidates[idx]);
      temp.pop_back();
      solve(idx+1,n,candidates,ans,temp,target);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        int n = candidates.size();
        vector<vector<int>>ans;
        vector<int> temp;
        solve(0,n,candidates,ans,temp, target);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public void solve(int[] candidates, int idx, int target, List<List<Integer>>ans, List<Integer> temp){
        if(target == 0){
            ans.add(new ArrayList<>(temp));
            return;
        }
        if(target < 0){
            return;
        }
        if(idx >= candidates.length){
            if(target == 0){
                ans.add(temp);
            }
            return;
        }
        temp.add(candidates[idx]);
        solve(candidates,idx,target - candidates[idx], ans, temp);
        temp.removeLast();
        solve(candidates, idx+1, target, ans, temp);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>>ans = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        int idx = 0;
        solve(candidates,idx, target,ans,temp );
        return ans;
    }
}